File: android/app/src/main/kotlin/com/example/ptchampion/domain/exercise/analyzers/PushupAnalyzer.kt
package com.example.ptchampion.domain.exercise.analyzers

import android.util.Log
import com.example.ptchampion.domain.exercise.AnalysisResult
import com.example.ptchampion.domain.exercise.ExerciseAnalyzer
import com.example.ptchampion.domain.exercise.ExerciseState
import com.example.ptchampion.domain.exercise.utils.AngleCalculator
import com.example.ptchampion.domain.exercise.utils.PoseLandmark
import com.example.ptchampion.domain.exercise.utils.PoseResultExtensions.getLandmark
import com.example.ptchampion.domain.exercise.utils.PoseResultExtensions.isLandmarkVisible
import com.example.ptchampion.domain.exercise.utils.PoseResultExtensions.areAllLandmarksVisible
import com.example.ptchampion.domain.exercise.utils.PoseResultExtensions.getAverageConfidence
import com.example.ptchampion.posedetection.PoseLandmarkerHelper
import com.google.mediapipe.tasks.components.containers.NormalizedLandmark
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarkerResult
import kotlin.math.abs
import kotlin.math.min
import kotlin.math.max
import java.util.Optional
import kotlin.math.sqrt
import kotlin.math.acos
import kotlin.math.pow

/**
 * Analyzer for push-ups according to US Army APFT standards.
 * 
 * Army standards require:
 * - Body forms a generally straight line from head to heels
 * - Starting position has fully extended arms
 * - Lowering until upper arms are at least parallel to the ground
 * - No arching of the back or raising the buttocks
 * - Movement as a single unit
 */
class PushupAnalyzer : ExerciseAnalyzer {

    companion object {
        private const val TAG = "PushupAnalyzer"
        
        // Constants for APFT standard form checking
        private const val MIN_ELBOW_ANGLE_THRESHOLD = 90f  // Upper arms must be parallel to ground
        private const val FULL_EXTENSION_THRESHOLD = 160f  // Arms nearly straight at top
        private const val SHOULDER_ALIGNMENT_THRESHOLD = 0.10f // X-axis alignment tolerance
        private const val HIP_SAG_THRESHOLD = 0.10f // Hip sagging threshold
        private const val HIP_PIKE_THRESHOLD = 0.12f // Hip raising/piking threshold
        private const val BODY_ALIGNMENT_THRESHOLD = 0.15f // Shoulders-hips-ankles alignment
        private const val REQUIRED_VISIBILITY = 0.5f       // Minimum visibility for landmarks
        
        // Key landmarks required for proper pushup analysis
        private val KEY_LANDMARKS = listOf(
            PoseLandmark.LEFT_SHOULDER,
            PoseLandmark.RIGHT_SHOULDER,
            PoseLandmark.LEFT_ELBOW,
            PoseLandmark.RIGHT_ELBOW,
            PoseLandmark.LEFT_WRIST,
            PoseLandmark.RIGHT_WRIST,
            PoseLandmark.LEFT_HIP,
            PoseLandmark.RIGHT_HIP,
            PoseLandmark.LEFT_KNEE,
            PoseLandmark.RIGHT_KNEE,
            PoseLandmark.LEFT_ANKLE,
            PoseLandmark.RIGHT_ANKLE
        )
    }

    private var repCount = 0
    private var currentState = ExerciseState.IDLE
    private var previousState = ExerciseState.IDLE
    private var minElbowAngle = 180f  // Track minimum angle during rep for depth check
    private var maxHipDeviation = 0f  // Track maximum hip deviation from straight line
    private val formIssues = mutableListOf<String>()
    
    // Store previous valid positions for movement analysis
    private var lastValidShoulderY = 0f
    private var lastValidHipY = 0f
    private var lastValidElbowAngle = 180f

    override fun analyze(resultBundle: PoseLandmarkerHelper.ResultBundle): AnalysisResult {
        formIssues.clear()
        
        // Extract MediaPipe results
        val results = resultBundle.results
        
        // Ensure we have landmarks
        if (results.landmarks().isEmpty()) {
            return AnalysisResult(
                repCount = repCount,
                feedback = "No pose detected",
                state = ExerciseState.INVALID,
                confidence = 0f,
                formScore = 0.0
            )
        }
        
        // Get the first person's landmarks
        val landmarks = results.landmarks()[0]
        
        // Check if we have necessary landmarks with good visibility
        if (!areKeyLandmarksVisible(landmarks)) {
            return AnalysisResult(
                repCount = repCount,
                feedback = "Position your full body in frame",
                state = ExerciseState.INVALID,
                confidence = getAverageConfidence(landmarks),
                formScore = 0.0
            )
        }

        try {
            // Extract key landmarks for analysis
            val leftShoulder = landmarks[PoseLandmark.LEFT_SHOULDER]
            val rightShoulder = landmarks[PoseLandmark.RIGHT_SHOULDER]
            val leftElbow = landmarks[PoseLandmark.LEFT_ELBOW]
            val rightElbow = landmarks[PoseLandmark.RIGHT_ELBOW]
            val leftWrist = landmarks[PoseLandmark.LEFT_WRIST]
            val rightWrist = landmarks[PoseLandmark.RIGHT_WRIST]
            val leftHip = landmarks[PoseLandmark.LEFT_HIP]
            val rightHip = landmarks[PoseLandmark.RIGHT_HIP]
            val leftKnee = landmarks[PoseLandmark.LEFT_KNEE]
            val rightKnee = landmarks[PoseLandmark.RIGHT_KNEE]
            val leftAnkle = landmarks[PoseLandmark.LEFT_ANKLE]
            val rightAnkle = landmarks[PoseLandmark.RIGHT_ANKLE]

            // Calculate joint angles
            val leftElbowAngle = calculateAngle(leftShoulder, leftElbow, leftWrist)
            val rightElbowAngle = calculateAngle(rightShoulder, rightElbow, rightWrist)
            val avgElbowAngle = (leftElbowAngle + rightElbowAngle) / 2

            // Store current position data
            val avgShoulderY = (leftShoulder.y() + rightShoulder.y()) / 2
            val avgHipY = (leftHip.y() + rightHip.y()) / 2
            val avgAnkleY = (leftAnkle.y() + rightAnkle.y()) / 2
            
            // Update position tracking
            if (avgElbowAngle > 0 && avgShoulderY > 0 && avgHipY > 0) {
                lastValidShoulderY = avgShoulderY
                lastValidHipY = avgHipY
                lastValidElbowAngle = avgElbowAngle
            }

            // Update minimum elbow angle during a rep
            if (currentState == ExerciseState.DOWN || 
                (currentState == ExerciseState.STARTING && previousState == ExerciseState.UP)) {
                minElbowAngle = min(minElbowAngle, avgElbowAngle)
            }

            // APFT Form checks
            
            // 1. Body straight line check (shoulders-hips-ankles)
            checkBodyAlignment(leftShoulder, rightShoulder, leftHip, rightHip, leftAnkle, rightAnkle)
            
            // 2. Hip sagging check
            checkHipSag(leftShoulder, rightShoulder, leftHip, rightHip, leftAnkle, rightAnkle)
            
            // 3. Hip piking check (butt raising too high)
            checkHipPike(leftShoulder, rightShoulder, leftHip, rightHip, leftAnkle, rightAnkle)
            
            // 4. Shoulder alignment check (level shoulders)
            checkShoulderAlignment(leftShoulder, rightShoulder)
            
            // 5. Proper arm movement check
            checkArmMovement(avgElbowAngle)

            // State machine logic to track rep phases
            previousState = currentState
            currentState = determineState(avgElbowAngle)

            // Rep counting logic - only count when proper form is maintained
            if (previousState == ExerciseState.DOWN && currentState == ExerciseState.UP) {
                // Check depth at bottom of pushup (did they go low enough?)
                if (minElbowAngle > MIN_ELBOW_ANGLE_THRESHOLD) {
                    formIssues.add("Not low enough (arms weren't parallel to ground)")
                } else {
                    // Only count the rep if it meets depth requirement
                    repCount++
                }
                
                // Reset tracking variables for next rep
                minElbowAngle = 180f
                maxHipDeviation = 0f
            }

            // Handle transition from IDLE or INVALID to STARTING
            if ((previousState == ExerciseState.IDLE || previousState == ExerciseState.INVALID) &&
                (currentState == ExerciseState.UP || currentState == ExerciseState.DOWN)) {
                currentState = ExerciseState.STARTING
            }

            // Generate feedback message
            val feedback = when {
                formIssues.isNotEmpty() -> formIssues.joinToString(". ")
                currentState == ExerciseState.UP -> "Good form, upper position"
                currentState == ExerciseState.DOWN -> "Good form, lower position"
                currentState == ExerciseState.STARTING -> "Starting position OK"
                else -> null
            }

            // Calculate overall form score
            val formScore = calculateFormScore(
                minElbowAngleAchieved = minElbowAngle,
                bodyDeviations = maxHipDeviation,
                shoulders = Pair(leftShoulder, rightShoulder),
                hips = Pair(leftHip, rightHip)
            )

            // Calculate confidence from landmark visibility
            val confidence = getAverageConfidence(landmarks)

            return AnalysisResult(
                repCount = repCount,
                feedback = feedback,
                state = currentState,
                confidence = confidence,
                formScore = formScore
            )
        } catch (e: Exception) {
            Log.e(TAG, "Error analyzing pose: ${e.message}", e)
            return AnalysisResult(
                repCount = repCount,
                feedback = "Error during analysis",
                state = ExerciseState.INVALID,
                confidence = 0f,
                formScore = 0.0
            )
        }
    }

    /**
     * Determines the current push-up state based on the elbow angle.
     */
    private fun determineState(elbowAngle: Float): ExerciseState {
        return when {
            elbowAngle < MIN_ELBOW_ANGLE_THRESHOLD * 1.1 -> ExerciseState.DOWN // Low position
            elbowAngle > FULL_EXTENSION_THRESHOLD * 0.95 -> ExerciseState.UP // High position
            currentState == ExerciseState.IDLE -> ExerciseState.STARTING
            else -> currentState // Maintain current state during transition
        }
    }

    /**
     * Checks if the body forms a straight line from shoulders to ankles.
     * APFT requires the body to move as a unit without sagging or piking.
     */
    private fun checkBodyAlignment(
        leftShoulder: NormalizedLandmark,
        rightShoulder: NormalizedLandmark,
        leftHip: NormalizedLandmark,
        rightHip: NormalizedLandmark,
        leftAnkle: NormalizedLandmark,
        rightAnkle: NormalizedLandmark
    ) {
        // Calculate average y-coordinates (vertical position)
        val avgShoulderY = (leftShoulder.y() + rightShoulder.y()) / 2
        val avgHipY = (leftHip.y() + rightHip.y()) / 2
        val avgAnkleY = (leftAnkle.y() + rightAnkle.y()) / 2
        
        // In a perfect plank position, shoulders, hips and ankles should form a straight line
        // We'll use a linear approximation and check deviation of hips from that line
        
        // Calculate expected hip position if on straight line between shoulders and ankles
        val shoulderToAnkleRatio = 0.5 // Hips should be approximately halfway
        val expectedHipY = avgShoulderY + (avgAnkleY - avgShoulderY) * shoulderToAnkleRatio
        
        // Calculate actual deviation
        val hipDeviation = abs(avgHipY - expectedHipY)
        
        // Track maximum hip deviation during the rep
        maxHipDeviation = kotlin.math.max(maxHipDeviation as Float, hipDeviation as Float)
        
        // Check if deviation exceeds threshold
        if (hipDeviation > BODY_ALIGNMENT_THRESHOLD) {
            if (avgHipY > expectedHipY) {
                formIssues.add("Keep your body straight, hips are sagging")
            } else {
                formIssues.add("Keep your body straight, hips are too high")
            }
        }
    }

    /**
     * Specifically checks for hip sagging (common issue).
     * In APFT, the torso must be kept rigid and move as a unit.
     */
    private fun checkHipSag(
        leftShoulder: NormalizedLandmark,
        rightShoulder: NormalizedLandmark,
        leftHip: NormalizedLandmark,
        rightHip: NormalizedLandmark,
        leftAnkle: NormalizedLandmark,
        rightAnkle: NormalizedLandmark
    ) {
        val avgShoulderX = (leftShoulder.x() + rightShoulder.x()) / 2
        val avgShoulderY = (leftShoulder.y() + rightShoulder.y()) / 2
        val avgShoulderZ = (leftShoulder.z() + rightShoulder.z()) / 2

        val avgHipX = (leftHip.x() + rightHip.x()) / 2
        val avgHipY = (leftHip.y() + rightHip.y()) / 2
        val avgHipZ = (leftHip.z() + rightHip.z()) / 2
        
        val avgAnkleX = (leftAnkle.x() + rightAnkle.x()) / 2
        val avgAnkleY = (leftAnkle.y() + rightAnkle.y()) / 2
        val avgAnkleZ = (leftAnkle.z() + rightAnkle.z()) / 2
        
        // Manual angle calculation using averaged coordinates
        val v1x = avgShoulderX - avgHipX
        val v1y = avgShoulderY - avgHipY
        val v1z = avgShoulderZ - avgHipZ
        val v2x = avgAnkleX - avgHipX
        val v2y = avgAnkleY - avgHipY
        val v2z = avgAnkleZ - avgHipZ
        val dotProduct = v1x * v2x + v1y * v2y + v1z * v2z
        val v1Mag = sqrt(v1x.pow(2) + v1y.pow(2) + v1z.pow(2))
        val v2Mag = sqrt(v2x.pow(2) + v2y.pow(2) + v2z.pow(2))
        
        val hipAngle = if (v1Mag > 0.0001f && v2Mag > 0.0001f) {
            val cosTheta = (dotProduct / (v1Mag * v2Mag)).coerceIn(-1.0f, 1.0f)
            Math.toDegrees(acos(cosTheta).toDouble()).toFloat()
        } else {
            0f // Default angle if magnitude is near zero
        }
        
        // If angle is significantly less than 180, hips are sagging
        if (hipAngle < 160 && avgHipY > (avgShoulderY + avgAnkleY) / 2) {
            formIssues.add("Keep your core tight, hips are sagging")
        }
    }

    /**
     * Checks for hip piking (raising the butt too high).
     * APFT requires the body to maintain a straight line.
     */
    private fun checkHipPike(
        leftShoulder: NormalizedLandmark,
        rightShoulder: NormalizedLandmark,
        leftHip: NormalizedLandmark,
        rightHip: NormalizedLandmark,
        leftAnkle: NormalizedLandmark,
        rightAnkle: NormalizedLandmark
    ) {
        val avgShoulderX = (leftShoulder.x() + rightShoulder.x()) / 2
        val avgShoulderY = (leftShoulder.y() + rightShoulder.y()) / 2
        val avgShoulderZ = (leftShoulder.z() + rightShoulder.z()) / 2

        val avgHipX = (leftHip.x() + rightHip.x()) / 2
        val avgHipY = (leftHip.y() + rightHip.y()) / 2
        val avgHipZ = (leftHip.z() + rightHip.z()) / 2

        val avgAnkleX = (leftAnkle.x() + rightAnkle.x()) / 2
        val avgAnkleY = (leftAnkle.y() + rightAnkle.y()) / 2
        val avgAnkleZ = (leftAnkle.z() + rightAnkle.z()) / 2
        
        // Manual angle calculation using averaged coordinates
        val v1x = avgShoulderX - avgHipX
        val v1y = avgShoulderY - avgHipY
        val v1z = avgShoulderZ - avgHipZ
        val v2x = avgAnkleX - avgHipX
        val v2y = avgAnkleY - avgHipY
        val v2z = avgAnkleZ - avgHipZ
        val dotProduct = v1x * v2x + v1y * v2y + v1z * v2z
        val v1Mag = sqrt(v1x.pow(2) + v1y.pow(2) + v1z.pow(2))
        val v2Mag = sqrt(v2x.pow(2) + v2y.pow(2) + v2z.pow(2))

        val hipAngle = if (v1Mag > 0.0001f && v2Mag > 0.0001f) {
            val cosTheta = (dotProduct / (v1Mag * v2Mag)).coerceIn(-1.0f, 1.0f)
            Math.toDegrees(acos(cosTheta).toDouble()).toFloat()
        } else {
            0f // Default angle if magnitude is near zero
        }
        
        // If angle is significantly more than 180, hips are piked (butt too high)
        if (hipAngle > 200 && avgHipY < (avgShoulderY + avgAnkleY) / 2) {
            formIssues.add("Lower your hips, buttocks are too high")
        }
    }

    /**
     * Checks if shoulders are level (not tilted).
     * APFT requires balanced, controlled movement.
     */
    private fun checkShoulderAlignment(
        leftShoulder: NormalizedLandmark,
        rightShoulder: NormalizedLandmark
    ) {
        val shoulderTilt = abs(leftShoulder.y() - rightShoulder.y())
        if (shoulderTilt > SHOULDER_ALIGNMENT_THRESHOLD) {
            formIssues.add("Keep shoulders level")
        }
    }

    /**
     * Monitors arm movement to ensure proper lowering and raising.
     * APFT requires controlled movement.
     */
    private fun checkArmMovement(currentElbowAngle: Float) {
        // Check for sudden changes in elbow angle which might indicate jerky movements
        val angleDelta = abs(currentElbowAngle - lastValidElbowAngle)
        
        // Only check when in motion (not at extremes)
        if (currentState != ExerciseState.IDLE && 
            currentElbowAngle > MIN_ELBOW_ANGLE_THRESHOLD && 
            currentElbowAngle < FULL_EXTENSION_THRESHOLD) {
            
            // Large sudden changes might indicate jerky motion, not controlled
            if (angleDelta > 30) {
                formIssues.add("Keep movements slow and controlled")
            }
        }
    }

    /**
     * Calculates form score based on APFT standards.
     * Returns 0-100 where 100 is perfect form.
     */
    private fun calculateFormScore(
        minElbowAngleAchieved: Float,
        bodyDeviations: Float,
        shoulders: Pair<NormalizedLandmark, NormalizedLandmark>,
        hips: Pair<NormalizedLandmark, NormalizedLandmark>
    ): Double {
        var score = 100.0
        
        // Penalty for insufficient depth (most critical for APFT)
        if (minElbowAngleAchieved > MIN_ELBOW_ANGLE_THRESHOLD) {
            val depthPenalty = ((minElbowAngleAchieved - MIN_ELBOW_ANGLE_THRESHOLD) * 1.5).toInt()
                .coerceAtMost(50) // Major penalty - up to 50 points
            score -= depthPenalty.toDouble()
        }
        
        // Penalty for body alignment issues
        val alignmentPenalty = (bodyDeviations * 300).toInt().coerceAtMost(40)
        score -= alignmentPenalty.toDouble()
        
        // Penalty for shoulder alignment issues
        val shoulderTilt = abs(shoulders.first.y() - shoulders.second.y())
        if (shoulderTilt > SHOULDER_ALIGNMENT_THRESHOLD) {
            val tiltPenalty = (shoulderTilt * 100).toInt().coerceAtMost(20)
            score -= tiltPenalty.toDouble()
        }
        
        return score.coerceIn(0.0, 100.0)
    }

    /**
     * Calculates the angle between three landmarks.
     */
    private fun calculateAngle(
        first: NormalizedLandmark,
        middle: NormalizedLandmark,
        last: NormalizedLandmark
    ): Float {
        // Calculate vectors from middle point
        val v1x = first.x() - middle.x()
        val v1y = first.y() - middle.y()
        val v1z = first.z() - middle.z()

        val v2x = last.x() - middle.x()
        val v2y = last.y() - middle.y()
        val v2z = last.z() - middle.z()

        // Calculate dot product
        val dotProduct = v1x * v2x + v1y * v2y + v1z * v2z

        // Calculate magnitudes
        val v1Mag = kotlin.math.sqrt(v1x * v1x + v1y * v1y + v1z * v1z)
        val v2Mag = kotlin.math.sqrt(v2x * v2x + v2y * v2y + v2z * v2z)

        // Handle potential division by zero
        if (v1Mag < 0.0001f || v2Mag < 0.0001f) {
            return 0f
        }

        // Calculate angle in radians, clamp to [-1, 1] to avoid NaN from acos
        val cosTheta = (dotProduct / (v1Mag * v2Mag)).coerceIn(-1.0f, 1.0f)
        val angleRad = kotlin.math.acos(cosTheta)

        // Convert to degrees
        return Math.toDegrees(angleRad.toDouble()).toFloat()
    }

    /**
     * Gets the average confidence across all landmarks.
     */
    private fun getAverageConfidence(landmarks: List<NormalizedLandmark>): Float {
        if (landmarks.isEmpty()) return 0f
        // Use the getVisibility extension function to safely handle Optional<Float> values
        val visibilities = landmarks.map { it.getVisibility() }
        return if (visibilities.isNotEmpty()) visibilities.average().toFloat() else 0f
    }

    /**
     * Checks if all key landmarks needed for pushup analysis are visible.
     */
    private fun areKeyLandmarksVisible(landmarks: List<NormalizedLandmark>): Boolean {
        if (landmarks.size < 33) return false // Full pose has 33 landmarks
        
        return KEY_LANDMARKS.all { landmarkIndex ->
            val landmark = landmarks.getOrNull(landmarkIndex)
            // Use the getVisibility extension function for consistent handling
            landmark != null && landmark.getVisibility() >= REQUIRED_VISIBILITY 
        }
    }

    override fun isValidPose(resultBundle: PoseLandmarkerHelper.ResultBundle): Boolean {
        if (resultBundle.results.landmarks().isEmpty()) return false
        return areKeyLandmarksVisible(resultBundle.results.landmarks()[0])
    }

    override fun start() {
        reset()
        currentState = ExerciseState.STARTING
    }

    override fun stop() {
        currentState = ExerciseState.FINISHED
    }

    override fun reset() {
        repCount = 0
        currentState = ExerciseState.IDLE
        previousState = ExerciseState.IDLE
        minElbowAngle = 180f
        maxHipDeviation = 0f
        lastValidShoulderY = 0f
        lastValidHipY = 0f
        lastValidElbowAngle = 180f
        formIssues.clear()
    }
} 

File: web/src/grading/PushupAnalyzer.ts
import { NormalizedLandmark } from '../lib/types';
import { PoseLandmarkIndex } from '../services/PoseLandmarkIndex';

/**
 * Interface for storing pushup form analysis results
 */
export interface PushupFormAnalysis {
  // Rep counting
  isDownPosition: boolean;
  isUpPosition: boolean;
  isValidRep: boolean;
  
  // Angle measurements
  leftElbowAngle: number;
  rightElbowAngle: number;
  bodyAlignmentAngle: number;
  
  // Form issues
  isBodySagging: boolean;
  isBodyPiking: boolean;
  isWorming: boolean;
  
  // Disqualifying breaks
  handsLiftedOff: boolean;
  feetLiftedOff: boolean;
  kneesTouchingGround: boolean;
  bodyTouchingGround: boolean;
  isPaused: boolean;
  
  // Debug info
  minElbowAngleDuringRep: number;
}

/**
 * Configuration options for push-up form analysis
 */
export interface PushupAnalyzerConfig {
  // Angle thresholds
  minElbowExtensionAngle: number; // Minimum angle considered "up" position (e.g., 160°)
  maxElbowFlexionAngle: number;   // Maximum angle considered "down" position (e.g., 90°)
  
  // Body alignment thresholds
  maxPikingAngle: number;         // Maximum angle before considered "piking" (e.g., 165°)
  minSaggingAngle: number;        // Minimum angle before considered "sagging" (e.g., 195°)
  
  // Movement thresholds
  wormingThreshold: number;       // Allowed difference in shoulder vs hip movement (in normalized units)
  groundTouchThreshold: number;   // Threshold for detecting body touching ground
  
  // Duration thresholds
  pauseThresholdMs: number;       // Time in ms before considered "paused" in position
}

// Default configuration values
const DEFAULT_CONFIG: PushupAnalyzerConfig = {
  minElbowExtensionAngle: 160,
  maxElbowFlexionAngle: 90,
  maxPikingAngle: 165,
  minSaggingAngle: 195,
  wormingThreshold: 0.03,
  groundTouchThreshold: 0.02,
  pauseThresholdMs: 2000,
};

/**
 * Class for analyzing push-up form using MediaPipe pose landmarks
 */
export class PushupAnalyzer {
  private config: PushupAnalyzerConfig;
  private prevLandmarks: NormalizedLandmark[] | null = null;
  private prevTimestamp: number = 0;
  private minElbowAngle: number = 180;
  private timeInCurrentPosition: number = 0;
  private lastPositionChangeTime: number = 0;
  
  constructor(config: Partial<PushupAnalyzerConfig> = {}) {
    this.config = { ...DEFAULT_CONFIG, ...config };
  }
  
  /**
   * Reset the analyzer state for a new set of push-ups
   */
  reset(): void {
    this.prevLandmarks = null;
    this.prevTimestamp = 0;
    this.minElbowAngle = 180;
    this.timeInCurrentPosition = 0;
    this.lastPositionChangeTime = 0;
  }
  
  /**
   * Analyze push-up form from pose landmarks
   * 
   * @param landmarks MediaPipe pose landmarks
   * @param timestamp Current timestamp in milliseconds
   * @returns Analysis results
   */
  analyzePushupForm(
    landmarks: NormalizedLandmark[],
    timestamp: number
  ): PushupFormAnalysis {
    // Calculate joint angles
    const leftElbowAngle = this.calculateAngle(
      landmarks[PoseLandmarkIndex.LEFT_SHOULDER],
      landmarks[PoseLandmarkIndex.LEFT_ELBOW],
      landmarks[PoseLandmarkIndex.LEFT_WRIST]
    );
    
    const rightElbowAngle = this.calculateAngle(
      landmarks[PoseLandmarkIndex.RIGHT_SHOULDER],
      landmarks[PoseLandmarkIndex.RIGHT_ELBOW],
      landmarks[PoseLandmarkIndex.RIGHT_WRIST]
    );
    
    // Calculate body alignment using midpoints
    const midShoulder = this.getMidpoint(
      landmarks[PoseLandmarkIndex.LEFT_SHOULDER],
      landmarks[PoseLandmarkIndex.RIGHT_SHOULDER]
    );
    
    const midHip = this.getMidpoint(
      landmarks[PoseLandmarkIndex.LEFT_HIP],
      landmarks[PoseLandmarkIndex.RIGHT_HIP]
    );
    
    const midAnkle = this.getMidpoint(
      landmarks[PoseLandmarkIndex.LEFT_ANKLE],
      landmarks[PoseLandmarkIndex.RIGHT_ANKLE]
    );
    
    // Calculate angle to check body alignment (shoulder-hip-ankle)
    const bodyAlignmentAngle = this.calculateAngle(midShoulder, midHip, midAnkle);
    
    // Check up and down positions
    const currentElbowAngle = Math.min(leftElbowAngle, rightElbowAngle);
    const isDownPosition = currentElbowAngle <= this.config.maxElbowFlexionAngle;
    const isUpPosition = currentElbowAngle >= this.config.minElbowExtensionAngle;
    
    // Track minimum elbow angle during the rep
    if (currentElbowAngle < this.minElbowAngle) {
      this.minElbowAngle = currentElbowAngle;
    }
    
    // Check body alignment issues
    const isBodySagging = bodyAlignmentAngle > this.config.minSaggingAngle;
    const isBodyPiking = bodyAlignmentAngle < this.config.maxPikingAngle;
    
    // Detect worming (shoulders and hips not moving in sync)
    let isWorming = false;
    if (this.prevLandmarks) {
      const shoulderMovement = Math.abs(midShoulder.y - this.getMidpoint(
        this.prevLandmarks[PoseLandmarkIndex.LEFT_SHOULDER],
        this.prevLandmarks[PoseLandmarkIndex.RIGHT_SHOULDER]
      ).y);
      
      const hipMovement = Math.abs(midHip.y - this.getMidpoint(
        this.prevLandmarks[PoseLandmarkIndex.LEFT_HIP],
        this.prevLandmarks[PoseLandmarkIndex.RIGHT_HIP]
      ).y);
      
      isWorming = Math.abs(shoulderMovement - hipMovement) > this.config.wormingThreshold;
    }
    
    // Detect disqualifying breaks
    const handsLiftedOff = this.detectLiftOff(
      landmarks[PoseLandmarkIndex.LEFT_WRIST],
      landmarks[PoseLandmarkIndex.RIGHT_WRIST],
      this.prevLandmarks ? [
        this.prevLandmarks[PoseLandmarkIndex.LEFT_WRIST],
        this.prevLandmarks[PoseLandmarkIndex.RIGHT_WRIST]
      ] : null
    );
    
    const feetLiftedOff = this.detectLiftOff(
      landmarks[PoseLandmarkIndex.LEFT_ANKLE],
      landmarks[PoseLandmarkIndex.RIGHT_ANKLE],
      this.prevLandmarks ? [
        this.prevLandmarks[PoseLandmarkIndex.LEFT_ANKLE],
        this.prevLandmarks[PoseLandmarkIndex.RIGHT_ANKLE]
      ] : null
    );
    
    // Detect knees touching ground
    const leftKneeY = landmarks[PoseLandmarkIndex.LEFT_KNEE].y;
    const rightKneeY = landmarks[PoseLandmarkIndex.RIGHT_KNEE].y;
    const leftHipY = landmarks[PoseLandmarkIndex.LEFT_HIP].y;
    const rightHipY = landmarks[PoseLandmarkIndex.RIGHT_HIP].y;
    
    const kneesTouchingGround = Math.abs(leftKneeY - leftHipY) < this.config.groundTouchThreshold || 
                               Math.abs(rightKneeY - rightHipY) < this.config.groundTouchThreshold;
    
    // Detect entire body touching ground
    const bodyTouchingGround = Math.abs(midShoulder.y - midHip.y) < this.config.groundTouchThreshold;
    
    // Check for pausing in position
    let isPaused = false;
    if (this.prevTimestamp > 0) {
      // Update time in current position
      const timeDiff = timestamp - this.prevTimestamp;
      
      if ((this.prevLandmarks && isUpPosition) || (this.prevLandmarks && isDownPosition)) {
        this.timeInCurrentPosition += timeDiff;
      } else {
        this.timeInCurrentPosition = 0;
        this.lastPositionChangeTime = timestamp;
      }
      
      isPaused = this.timeInCurrentPosition > this.config.pauseThresholdMs;
    }
    
    // Determine if the current rep is valid
    const isValidRep = !isBodySagging && 
                      !isBodyPiking && 
                      !isWorming && 
                      !handsLiftedOff && 
                      !feetLiftedOff && 
                      !kneesTouchingGround && 
                      !bodyTouchingGround && 
                      !isPaused && 
                      this.minElbowAngle <= this.config.maxElbowFlexionAngle;
    
    // Update state for next analysis
    this.prevLandmarks = [...landmarks];
    this.prevTimestamp = timestamp;
    
    // Return analysis results
    return {
      isDownPosition,
      isUpPosition,
      isValidRep,
      leftElbowAngle,
      rightElbowAngle,
      bodyAlignmentAngle,
      isBodySagging,
      isBodyPiking,
      isWorming,
      handsLiftedOff,
      feetLiftedOff,
      kneesTouchingGround,
      bodyTouchingGround,
      isPaused,
      minElbowAngleDuringRep: this.minElbowAngle
    };
  }
  
  /**
   * Calculate the angle between three points (in degrees)
   */
  private calculateAngle(
    a: NormalizedLandmark,
    b: NormalizedLandmark,
    c: NormalizedLandmark
  ): number {
    // Create vectors
    const vectorBA = {
      x: a.x - b.x,
      y: a.y - b.y,
      z: a.z - b.z
    };
    
    const vectorBC = {
      x: c.x - b.x,
      y: c.y - b.y,
      z: c.z - b.z
    };
    
    // Calculate dot product
    const dotProduct = vectorBA.x * vectorBC.x + vectorBA.y * vectorBC.y + vectorBA.z * vectorBC.z;
    
    // Calculate magnitudes
    const magnitudeBA = Math.sqrt(vectorBA.x * vectorBA.x + vectorBA.y * vectorBA.y + vectorBA.z * vectorBA.z);
    const magnitudeBC = Math.sqrt(vectorBC.x * vectorBC.x + vectorBC.y * vectorBC.y + vectorBC.z * vectorBC.z);
    
    // Calculate angle in radians
    const angleRad = Math.acos(dotProduct / (magnitudeBA * magnitudeBC));
    
    // Convert to degrees
    const angleDeg = angleRad * (180 / Math.PI);
    
    return angleDeg;
  }
  
  /**
   * Calculate midpoint between two landmarks
   */
  private getMidpoint(a: NormalizedLandmark, b: NormalizedLandmark): NormalizedLandmark {
    return {
      x: (a.x + b.x) / 2,
      y: (a.y + b.y) / 2,
      z: (a.z + b.z) / 2,
      visibility: Math.min(a.visibility || 0, b.visibility || 0)
    };
  }
  
  /**
   * Detect if landmarks have lifted off from their position
   */
  private detectLiftOff(
    leftLandmark: NormalizedLandmark,
    rightLandmark: NormalizedLandmark,
    prevLandmarks: NormalizedLandmark[] | null
  ): boolean {
    // If no previous landmarks, can't detect lift-off
    if (!prevLandmarks) return false;
    
    // Check for significant change in Y coordinate (vertical movement)
    const leftDiffY = Math.abs(leftLandmark.y - prevLandmarks[0].y);
    const rightDiffY = Math.abs(rightLandmark.y - prevLandmarks[1].y);
    
    // Detect if either landmark has moved significantly
    return leftDiffY > 0.03 || rightDiffY > 0.03;
  }
  
  /**
   * Reset minimum elbow angle tracked during a rep
   */
  resetMinElbowAngle(): void {
    this.minElbowAngle = 180;
  }
} 